--- library/error.c	2014-07-11 12:28:56.000000000 +0300
+++ library/error.c	2014-10-02 10:46:09.000000000 +0300
@@ -121,6 +121,10 @@
 #include "polarssl/pk.h"
 #endif

+#if defined(POLARSSL_ECIES_C)
+#include "polarssl/ecies.h"
+#endif
+
 #if defined(POLARSSL_PKCS12_C)
 #include "polarssl/pkcs12.h"
 #endif
@@ -308,6 +312,15 @@
             snprintf( buf, buflen, "PK - Unavailable feature, e.g. RSA disabled for RSA key" );
 #endif /* POLARSSL_PK_C */

+#if defined(POLARSSL_ECIES_C)
+        if( use_ret == -(POLARSSL_ERR_ECIES_BAD_INPUT_DATA) )
+            snprintf( buf, buflen, "ECIES - Bad input parameters to function" );
+        if( use_ret == -(POLARSSL_ERR_ECIES_OUTPUT_TOO_SMALL) )
+            snprintf( buf, buflen, "ECIES - Buffer is too small to write to" );
+        if( use_ret == -(POLARSSL_ERR_ECIES_MALFORMED_DATA) )
+            snprintf( buf, buflen, "ECIES - Encrypted data is malformed" );
+#endif /* POLARSSL_ECIES_C */
+
 #if defined(POLARSSL_PKCS12_C)
         if( use_ret == -(POLARSSL_ERR_PKCS12_BAD_INPUT_DATA) )
             snprintf( buf, buflen, "PKCS12 - Bad input parameters to function" );
