--- library/error.c	2014-07-11 12:28:56.000000000 +0300
+++ library/error.c	2015-02-09 20:01:12.000000000 +0200
@@ -175,6 +175,22 @@
 #include "polarssl/xtea.h"
 #endif
 
+#if defined(POLARSSL_ECIES_C)
+#include "polarssl/ecies.h"
+#endif
+
+#if defined(POLARSSL_KDF_C)
+#include "polarssl/kdf.h"
+#endif
+
+#if defined(POLARSSL_KDF1_C)
+#include "polarssl/kdf1.h"
+#endif
+
+#if defined(POLARSSL_KDF2_C)
+#include "polarssl/kdf2.h"
+#endif
+
 
 #include <string.h>
 
@@ -492,6 +508,22 @@
 #endif /* POLARSSL_X509_USE,X509_CREATE_C */
         // END generated code
 
+        // START custom added code
+#if defined(POLARSSL_ECIES_C)
+        if( use_ret == -(POLARSSL_ERR_ECIES_BAD_INPUT_DATA) )
+            snprintf( buf, buflen, "ECIES - Bad input parameters to function" );
+        if( use_ret == -(POLARSSL_ERR_ECIES_OUTPUT_TOO_SMALL) )
+            snprintf( buf, buflen, "ECIES - Buffer is too small to write to" );
+        if( use_ret == -(POLARSSL_ERR_ECIES_MALFORMED_DATA) )
+            snprintf( buf, buflen, "ECIES - Encrypted data is malformed" );
+#endif /* POLARSSL_ECIES_C */
+
+#if defined(POLARSSL_KDF_C)
+        if( use_ret == -(POLARSSL_ERR_KDF_BAD_INPUT_DATA) )
+            snprintf( buf, buflen, "KDF - Bad input parameters to function" );
+#endif /* POLARSSL_KDF_C */
+        // END custom added code
+
         if( strlen( buf ) == 0 )
             snprintf( buf, buflen, "UNKNOWN ERROR CODE (%04X)", use_ret );
     }
@@ -728,6 +760,18 @@
 #endif /* POLARSSL_XTEA_C */
     // END generated code
 
+    // START custom added code
+#if defined(POLARSSL_KDF1_C)
+    if( use_ret == -(POLARSSL_ERR_KDF1_BAD_INPUT_DATA) )
+        snprintf( buf, buflen, "KDF1 - Bad input parameters to function" );
+#endif /* POLARSSL_KDF1_C */
+
+#if defined(POLARSSL_KDF2_C)
+    if( use_ret == -(POLARSSL_ERR_KDF2_BAD_INPUT_DATA) )
+        snprintf( buf, buflen, "KDF2 - Bad input parameters to function" );
+#endif /* POLARSSL_KDF2_C */
+    // END custom added code
+
     if( strlen( buf ) != 0 )
         return;
 
