--- include/polarssl/oid.h	2014-07-11 12:28:56.000000000 +0300
+++ include/polarssl/oid.h	2014-10-01 07:29:13.000000000 +0300
@@ -546,6 +546,18 @@
  * \return         0 if successful, or POLARSSL_ERR_OID_NOT_FOUND
  */
 int oid_get_cipher_alg( const asn1_buf *oid, cipher_type_t *cipher_alg );
+
+/**
+ * \brief          Translate cipher_type into encryption algorithm OID
+ *
+ * \param cipher_alg   cipher algorithm
+ * \param oid          place to store ASN.1 OID string pointer
+ * \param olen         length of the OID
+ *
+ * \return         0 if successful, or POLARSSL_ERR_OID_NOT_FOUND
+ */
+int oid_get_oid_by_cipher_alg( cipher_type_t cipher_alg,
+                            const char **oid, size_t *olen );
 #endif /* POLARSSL_CIPHER_C */

 #if defined(POLARSSL_PKCS12_C)
@@ -561,6 +573,20 @@
  */
 int oid_get_pkcs12_pbe_alg( const asn1_buf *oid, md_type_t *md_alg,
                             cipher_type_t *cipher_alg );
+
+/**
+ * \brief          Translate md_type and cipher_type into
+ *                 PKCS#12 PBE algorithm OID
+ *
+ * \param md_alg   message digest algorithm
+ * \param pk_alg   public key algorithm
+ * \param oid      place to store ASN.1 OID string pointer
+ * \param olen     length of the OID
+ *
+ * \return         0 if successful, or POLARSSL_ERR_OID_NOT_FOUND
+ */
+int oid_get_oid_by_pkcs12_pbe_alg( cipher_type_t cipher_alg, md_type_t md_alg,
+                            const char **oid, size_t *olen );
 #endif /* POLARSSL_PKCS12_C */

 #ifdef __cplusplus
