--- library/oid.c	2014-07-11 12:28:56.000000000 +0300
+++ library/oid.c	2014-10-01 10:18:18.000000000 +0300
@@ -504,6 +504,7 @@

 FN_OID_TYPED_FROM_ASN1(oid_cipher_alg_t, cipher_alg, oid_cipher_alg);
 FN_OID_GET_ATTR1(oid_get_cipher_alg, oid_cipher_alg_t, cipher_alg, cipher_type_t, cipher_alg);
+FN_OID_GET_OID_BY_ATTR1(oid_get_oid_by_cipher_alg, oid_cipher_alg_t, oid_cipher_alg, cipher_type_t, cipher_alg);
 #endif /* POLARSSL_CIPHER_C */

 #if defined(POLARSSL_MD_C)
@@ -588,6 +589,7 @@

 FN_OID_TYPED_FROM_ASN1(oid_pkcs12_pbe_alg_t, pkcs12_pbe_alg, oid_pkcs12_pbe_alg);
 FN_OID_GET_ATTR2(oid_get_pkcs12_pbe_alg, oid_pkcs12_pbe_alg_t, pkcs12_pbe_alg, md_type_t, md_alg, cipher_type_t, cipher_alg);
+FN_OID_GET_OID_BY_ATTR2(oid_get_oid_by_pkcs12_pbe_alg, oid_pkcs12_pbe_alg_t, oid_pkcs12_pbe_alg, cipher_type_t, cipher_alg, md_type_t, md_alg);
 #endif /* POLARSSL_PKCS12_C */

 #if defined(_MSC_VER) && !defined snprintf && !defined(EFIX64) && \
